docs/madr/decisions/005-policy-matching.md

```diff
@@ -0,0 +1,647 @@
+mesh: mesh-1
+name: timeouts
+spec:
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
+    # timeouts configuration
+```
+
+"TargetRef" can attach policy to different objects:
+* Mesh
+* MeshSubset
+* Service
+* ServiceSubset
+* Proxy
+* MeshGatewayRoute 
+* HTTPRoute
+
+Traffic direction is expressed inside the "conf" section.
+
+If policy configures outbounds, then "conf" section contains "to" array:
+```yaml
+type: UpstreamTimeout
+mesh: mesh-1
+name: timeouts
+spec:
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
+    to:
+      - targetRef:
+          kind: Service
+          name: payments
+        connectTimeout: 10s
+      - targetRef:
+          kind: Service
+          name: web-api
+        connectTimeout: 20s
+```
+
+If policy configures inbounds, then "conf" section contains "from" array:
+```yaml
+type: MeshTrafficPermission
+mesh: mesh-1
+name: permissions
+spec:
+  targetRef:
+    kind: Service
+    name: backend
+  conf:
+    from:
+      - targetRef:
+          kind: MeshSubset
+          tags:
+            kuma.io/zone: us-east
+        action: ALLOW
+      - targetRef:
+          kind: ServiceSubset
+          name: backend
+          tags:
+            version: v2
+        action: ALLOW
+      - targetRef:
+          kind: Mesh 
+        action: DENY
+```
+
+### Applying
+
+#### Outbound policy
+
+All outbound policies contains a "to" array.
+Each element of the array has a "targetRef" that selects outbound
+and has an outbound configuration:
+
+```yaml
+to:
+  - targetRef: Mesh | Service | HTTPRoute
+    outboundConf: ...
+```
+
+If a data plane proxy has 3 outbounds:
+```yaml
+outbound:
+  - port: 8081
+    tags:
+      kuma.io/service: backend
+  - port: 8082
+    tags:
+      kuma.io/service: web-api
+  - port: 8083
+    tags:
+      kuma.io/service: payments
+```
+
+then policy can select a single "web-api" outbound:
+
+```yaml
+to:
+  - targetRef:
+      kind: Service
+      name: web-api
+    outboundConf: ... # conf for "web-api" outbound
+```
+
+policy can select all outbounds ("backend", "web-api" and "payments"):
+
+```yaml
+to:
+  - targetRef:
+      kind: Mesh
+    outboundConf: ... # conf for all outbounds
+```
+
+policy can select an HTTPRoute:
+
+```yaml
+to:
+  - targetRef:
+      kind: HTTPRoute
+      name: payments-get-info-v1-httproute
+    outboundConf: ... # conf for "/v1/getinfo/" route of "payments" outbound 
+```
+
+When several targetRefs select the same outbound, 
+outboundConfs are merged in the order they're presented in the "to" array:
+
+"backend" outbound is selected by 3 targetRefs
+
+```yaml
+to:
+  - targetRef: 
+      kind: Mesh
+    param1: value1
+  - targetRef: 
+      kind: Service
+      name: backend
+    param1: value2
+    param2: value3
+  - targetRef: 
+      kind: Mesh
+    param2: value4
+```
+
+resulting configuration for "backend" outbound is:
+
+```yaml
+param1: value1 # overrides 'value2' from 'targetRef{kind:Service,name:backend}' 
```

Comment:
Hmm, this feels backwards. Shouldn't later entries override earlier ones?

Reply:
> outboundConfs are merged in the order they're presented in the "to" array

I meant order from top to bottom, I can make it clearer

Reply:
I mean I would expect to have `param1: value2` then since the `Service` would override `Mesh`.

Reply:
No, it's actually simpler. Only order matters at this point. We merge `outboundConfs` without looking at `targetRefs`